<feat> Remove duplications when retrieving topk

sooyounlee opened the pull request:
- `base_retrieval.py`의 `retrieve` function에 문서 비교를 추가해서, 중복 문서일 경우 제외하는 코드를 추가했습니다.
- 미리 중복을 없앤 데이터를 사용하는 경우를 고려하여, `base_retrieval`에 `self.context_ids` 변수를 추가해서, `retrieve` function 호출 후 데이터가 return될 때 wiki text에 mapping된 `context_id`를 return하도록 변경했습니다. 
- `base_retrieval.py`의 `retrieve`에서 top-k를 뽑게 하기 위해 `base_retrieval.py`, `sparse_base.py`, `bm25.py`, `tfidf.py`, `dense_base.py`, `hybrid_base.py` 의 `get_relevant_doc_bulk`의 topk argument를 없애고, topk개 대신 모든 score와 index를 return하도록 함수를 변경했습니다.

sooyounlee committed:
<feat> add wiki data aggregation option

add wiki data aggregation option to the code

sooyounlee committed:
Merge branch 'main' of https://github.com/VumBleBot/odqa_baseline_code into sooyoun

sooyounlee committed:
<feat> Add wiki aggregation function

sooyounlee committed:
Merge branch 'main' of https://github.com/VumBleBot/odqa_baseline_code into sooyoun

sooyounlee committed:
Merge branch 'gunmo' of https://github.com/VumBleBot/odqa_baseline_code into sooyoun

sooyounlee committed:
<feat> remove duplications when retrieving topk

ggm1207 commented:
확인했습니다! 수연님 아까 논의 됐던 것과 달리 topk를 N 배수로 뽑는 것은 어떨까요? ㅠㅜ

아까는 생각이 안나서 말씀을 못 드렸는데 HybridRetrieval같은 경우가 sparse에서 top-k, dense에서 top-k 를 뽑아와서 새로운 score를 계산하는 방식이라서, topk 제한을 풀어버리면 계산량이 너무 많아질 것 같습니다!

아래는 관련 코드입니다!
```python
    # 새로운 Score를 계산하는 함수!!
    def _rank_fusion_by_hybrid(self, dense_hits, sparse_hits):
        ranks = []
        min_dense_score, min_sparse_score = min(dense_hits.values()), min(sparse_hits.values())

        for doc in set(dense_hits.keys()) | set(sparse_hits.keys()):
            if doc not in dense_hits:  # sparse 에만 들었을 때
                score = self.args.retriever.alpha * sparse_hits[doc] + min_dense_score
            elif doc not in sparse_hits:  # dense 에만 들었을 때
                score = self.args.retriever.alpha * min_sparse_score + dense_hits[doc]
            else:  # 둘 다 들었을 때
                score = self.args.retriever.alpha * sparse_hits[doc] + dense_hits[doc]

            ranks.append((doc, score))

        ranks = sorted(ranks, key=lambda x: x[1], reverse=True)[: self.args.retriever.topk]

        doc_index, doc_score = zip(*ranks)
        doc_index, doc_score = list(doc_index), list(doc_score)

        return doc_score, doc_index

    def get_relevant_doc_bulk(self, queries, topk):
        dense_scores, dense_indices = self.dense_retriever.get_relevant_doc_bulk(queries, topk)
        sparse_scores, sparse_indices = self.sparse_retriever.get_relevant_doc_bulk(queries, topk)

        doc_scores, doc_indices = [], []

        for idx, _ in enumerate(queries):
            scores_topk, indexs_topk = dense_scores[idx], dense_indices[idx]
            dense_hits = {indexs_topk[k]: scores_topk[k] for k in range(topk)}

            scores_topk, indexs_topk = sparse_scores[idx], sparse_indices[idx]
            sparse_hits = {indexs_topk[k]: scores_topk[k] for k in range(topk)}

            doc_score, doc_index = self._rank_fusion_by_hybrid(dense_hits, sparse_hits)

            doc_scores.append(doc_score)
            doc_indices.append(doc_index)

        return doc_scores, doc_indices
```

sooyounlee committed:
<fix> Fix the topk retrieval to retrieve 2*topk

sooyounlee commented:
> 확인했습니다! 수연님 아까 논의 됐던 것과 달리 topk를 N 배수로 뽑는 것은 어떨까요? ㅠㅜ
> 
> 아까는 생각이 안나서 말씀을 못 드렸는데 HybridRetrieval같은 경우가 sparse에서 top-k, dense에서 top-k 를 뽑아와서 새로운 score를 계산하는 방식이라서, topk 제한을 풀어버리면 계산량이 너무 많아질 것 같습니다!
> 
> 아래는 관련 코드입니다!
> 
> ```python
>     # 새로운 Score를 계산하는 함수!!
>     def _rank_fusion_by_hybrid(self, dense_hits, sparse_hits):
>         ranks = []
>         min_dense_score, min_sparse_score = min(dense_hits.values()), min(sparse_hits.values())
> 
>         for doc in set(dense_hits.keys()) | set(sparse_hits.keys()):
>             if doc not in dense_hits:  # sparse 에만 들었을 때
>                 score = self.args.retriever.alpha * sparse_hits[doc] + min_dense_score
>             elif doc not in sparse_hits:  # dense 에만 들었을 때
>                 score = self.args.retriever.alpha * min_sparse_score + dense_hits[doc]
>             else:  # 둘 다 들었을 때
>                 score = self.args.retriever.alpha * sparse_hits[doc] + dense_hits[doc]
> 
>             ranks.append((doc, score))
> 
>         ranks = sorted(ranks, key=lambda x: x[1], reverse=True)[: self.args.retriever.topk]
> 
>         doc_index, doc_score = zip(*ranks)
>         doc_index, doc_score = list(doc_index), list(doc_score)
> 
>         return doc_score, doc_index
> 
>     def get_relevant_doc_bulk(self, queries, topk):
>         dense_scores, dense_indices = self.dense_retriever.get_relevant_doc_bulk(queries, topk)
>         sparse_scores, sparse_indices = self.sparse_retriever.get_relevant_doc_bulk(queries, topk)
> 
>         doc_scores, doc_indices = [], []
> 
>         for idx, _ in enumerate(queries):
>             scores_topk, indexs_topk = dense_scores[idx], dense_indices[idx]
>             dense_hits = {indexs_topk[k]: scores_topk[k] for k in range(topk)}
> 
>             scores_topk, indexs_topk = sparse_scores[idx], sparse_indices[idx]
>             sparse_hits = {indexs_topk[k]: scores_topk[k] for k in range(topk)}
> 
>             doc_score, doc_index = self._rank_fusion_by_hybrid(dense_hits, sparse_hits)
> 
>             doc_scores.append(doc_score)
>             doc_indices.append(doc_index)
> 
>         return doc_scores, doc_indices
> ```

네! 안그래도 그냥 다 가져오면 어디에선가 계산이 많아지지 않을까 막연히 생각만 하고 있었는데 짚어주셔서 감사해요 :)
2배수 뽑아오는걸로 수정했습니다.

sooyounlee committed:
Merge branch 'main' of https://github.com/VumBleBot/odqa_baseline_code into sooyoun

ggm1207 commented on retrieval/base_retrieval.py:
```diff
@@ -26,23 +28,44 @@ def get_embedding(self):
         total = []
-        doc_scores, doc_indices = self.get_relevant_doc_bulk(query_or_dataset["question"], topk)
+        # 중복을 걸러내기 위해 topk를 2배수로 뽑습니다.
+        doc_scores, doc_indices = self.get_relevant_doc_bulk(query_or_dataset["question"], topk=2*topk)
```
topk 값이 3 ~ 5일 경우 중복이 많이 겹치게 되면 topk 개수를 못 채울 수도 있을 것 같다는 생각이 드는데 `min(30, topk*2)` 이런식으로 제한 값을 두는 건 어떨까요? :thinking:

ggm1207 commented on run_retrieval.py:
```diff
@@ -93,19 +94,18 @@ def train_retriever(args):
+                # if two texts overlaps more than 50%,
+                if fuzz.ratio(real, pred) > 50 and not qc_dict[question]:
                     qc_dict[question] = True
                     cur_cnt += 1
```
이 50이라는 기준이 어떻게 설정됐는지 궁금한데 이따 피어세션때 설명해주시면 좋을 것 같습니다!

sooyounlee committed:
<fix> Fixed topk document retrieval

- Add assert to ensure top-k document retrieval
- Increase fuzzy ratio from 50 to 65
- Changed topk to max(40 + topk, alpha*topk) and default alpha to be 3
